src/System.Private.CoreLib/src/System/Threading/Overlapped.cs

```diff
@@ -135,22 +136,100 @@ sealed internal unsafe class OverlappedData
             return AllocateNativeOverlapped();
         }
 
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        private extern NativeOverlapped* AllocateNativeOverlapped();
+        private unsafe NativeOverlapped* AllocateNativeOverlapped()
+        {
+            Debug.Assert(_pinnedData == null);
 
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal static extern void FreeNativeOverlapped(NativeOverlapped* nativeOverlappedPtr);
+            bool success = false;
+            try
+            {
+                if (_userObject != null)
+                {
+                    if (_userObject.GetType() == typeof(object[]))
+                    {
+                        object[] objArray = (object[])_userObject;
+
+                        _pinnedData = new GCHandle[objArray.Length];
+                        for (int i = 0; i < objArray.Length; i++)
+                        {
+                            _pinnedData[i] = GCHandle.Alloc(objArray[i], GCHandleType.Pinned);
+                        }
+                    }
+                    else
+                    {
+                        _pinnedData = new GCHandle[1];
+                        _pinnedData[0] = GCHandle.Alloc(_userObject, GCHandleType.Pinned);
+                    }
+                }
 
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal static extern OverlappedData GetOverlappedFromNative(NativeOverlapped* nativeOverlappedPtr);
+                NativeOverlapped* pNativeOverlapped = (NativeOverlapped*)Marshal.AllocHGlobal(sizeof(NativeOverlapped) + sizeof(GCHandle));
+                *(GCHandle*)(pNativeOverlapped + 1) = default(GCHandle);
+                _pNativeOverlapped = pNativeOverlapped;
+
+                _pNativeOverlapped->InternalLow = default;
+                _pNativeOverlapped->InternalHigh = default;
+                _pNativeOverlapped->OffsetLow = _offsetLow;
+                _pNativeOverlapped->OffsetHigh = _offsetHigh;
+                _pNativeOverlapped->EventHandle = _eventHandle;
+
+                *(GCHandle*)(_pNativeOverlapped + 1) = GCHandle.Alloc(this);
+
+                //if (ETW_EVENT_ENABLED(MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_Context, ThreadPoolIODequeue))
+                //    FireEtwThreadPoolIOPack(lpOverlapped, overlappedUNSAFE, GetClrInstanceId());
```

Comment:
Not sure how to translate this to managed code? I looked into `FrameworkEventSource`, but the definitions there are not really aligned with `ClrEtwAll.man`.

Reply:
The appropriate thing to do is to create an event that logs when an I/O packing happens as an event in FrameworkEventSource (as you guessed).  Yes, there is not particular correspondence, and the update may break causality reconstruction logic in PerfView (it will have to be updated to use the new event).   I can help with that.  (you can see the current use in https://github.com/Microsoft/perfview/blob/master/src/TraceEvent/Computers/ActivityComputer.cs#L158).  

But the first thing to do is to have event to replace it.  Fundamentally we want to link when async I/O is requested and when it completes (by some ID like the address of the NativeOverlapped structure, or anything else that is shared uniquely between the IO request and its completion).

Reply:
Ah, I just realised why the event is needed. I was always testing code that called both the `ThreadPoolIOPack` and `ThreadPoolIOEnqueue` events and not the native (file/socket) overlapped I/O where only `ThreadPoolIOPack` is triggered. I will look into it tomorrow to see how to add the event back.

Reply:
I gave it a whirl, but I am not happy with the result. Now `ThreadPoolIOEnqueue` and `ThreadPoolIOPack` appear in different providers. They can be correlated based on the parameters, but it's a bit suboptimal. Moreover, PerfView always picks the manifest for FrameworkEventSource from full NetFX instead of the CoreFX version, so the new event is not properly decoded.

Reply:
Would it be too bad to keep emitting the event using FCall?

Reply:
@jkotas Probably not. I have a prototype ready, but I didn't benchmark it yet. I also tried emitting the exact same event from managed code, but could not avoid getting some information from runtime and requiring FCall anyway.

Reply:
If you need to make a FCall anyway, it would be nice to simply log the event in native code, which avoids breaking compat.

Reply:
I need FCall if I try to log the event in the same format as before. I run some benchmarks, but the results were not very appealing and they were quite inconsistent (some runs were 2x as slow as others). I suspect there's something wrong either with the implementation or the benchmark... I'll look into it more.

Reply:
Apparently happens on the machine even for the unchanged CoreCLR, so I guess some power management issue :-/